src/queryHistory/queryHistoryProvider.ts

```diff
@@ -0,0 +1,208 @@
+        }
+        return this._queryHistoryNodes;
+    }
+
+    /**
+     *
+     */
+    public async showQueryHistoryCommandPalette(): Promise<void> {
+        const options = this._queryHistoryNodes.map(node => this._queryHistoryUI.convertToQuickPickItem(node));
+        let queryHistoryQuickPickItem = await this._queryHistoryUI.showQueryHistoryCommandPalette(options);
+        this.openQueryHistoryEntry(queryHistoryQuickPickItem.node, queryHistoryQuickPickItem.action ===
+            QueryHistoryAction.RunQueryHistoryAction);
+    }
+
+    /**
+     * Starts the history capture by changing the setting
+     * and changes context for menu actions
+     */
+    public async startQueryHistoryCapture(): Promise<void> {
+        await this._vscodeWrapper.setConfiguration(Constants.extensionConfigSectionName,
+            Constants.configEnableQueryHistoryCapture, true);
+        await this._vscodeWrapper.setContext(Constants.isQueryHistoryEnabled, true);
+    }
+
+    /**
+     * Pauses the history capture by changing the setting
+     * and changes context for menu actions
+     */
+    public async pauseQueryHistoryCapture(): Promise<void> {
+        await this._vscodeWrapper.setConfiguration(Constants.extensionConfigSectionName,
+            Constants.configEnableQueryHistoryCapture, false);
+        await this._vscodeWrapper.setContext(Constants.isQueryHistoryEnabled, false);
+    }
+
+    /**
+     * Opens a query history listing in a new query window
+     */
+    public async openQueryHistoryEntry(node: QueryHistoryNode, isExecute: boolean = false): Promise<void> {
+        const editor = await this._untitledSqlDocumentService.newQuery(node.queryString);
+        let uri = editor.document.uri.toString(true);
+        let title = path.basename(editor.document.fileName);
+        const queryUriPromise = new Deferred<boolean>();
+        let credentials = this._connectionManager.getConnectionInfo(node.ownerUri).credentials;
+        await this._connectionManager.connect(uri, credentials, queryUriPromise);
+        await queryUriPromise;
+        this._statusView.languageFlavorChanged(uri, Constants.mssqlProviderName);
+        this._statusView.sqlCmdModeChanged(uri, false);
+        if (isExecute) {
+            const queryPromise = new Deferred<boolean>();
+            await this._outputContentProvider.runQuery(this._statusView, uri, undefined, title, queryPromise);
+            await queryPromise;
+            await this._connectionManager.connectionStore.removeRecentlyUsed(<IConnectionProfile>credentials);
+        }
+    }
+
+    /**
+     * Deletes a query history entry for a URI
+     */
+    public deleteQueryHistoryEntry(node: QueryHistoryNode): void {
+        let index = this._queryHistoryNodes.findIndex(n => {
+            let historyNode = n as QueryHistoryNode;
+            return (historyNode.ownerUri === node.ownerUri) &&
+                historyNode.timeStamp === node.timeStamp;
```

Comment:
Still not sure why you're having this extra logic. Is the node object being returned not the same instance as the one in the nodes list? i.e. couldn't you just do n === historyNode? Using the timestamp is probably ok I guess but it seems more hacky than should be necessary

Reply:
I saw some weird behavior earlier, but I don't see it now. I'll remove the node itself.